core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala

```diff
@@ -372,12 +453,28 @@ object SparkSubmit {
       sysProps("spark.submit.pyFiles") = formattedPyFiles
     }
 
+    // NOTE: If we are using the REST gateway, we will use the original arguments directly.
+    // Since we mutate the values of some configs in this method, we must update the
+    // corresponding fields in the original SparkSubmitArguments to reflect these changes.
+    if (args.isStandaloneCluster) {
```

Comment:
What about moving this logic outside of this prepare environment function, and just deal with it when you know you are about to do a rest submission. There you have access to the sys props and original arguments.

Also, I really wonder if we can simplify this code. It's very complicated to trace down what is going on here.